Support for tagged measurements

chancefeick opened the pull request:
Add support for tagged measurements.

Tagged measurements consist of **tags**, which are a set of name=value tag pairs that describe the particular data stream. Tags are equivalent to **source** in the previous data model.
## Metrics

Introduces the following tagged metrics:

| Metric | Tags | Example |
| --- | --- | --- |
| `rails.request.total` | `controller`, `action`, `format` | controller=articles,action=show,format=json |
| `rails.request.status` | `status` | status=200 |
| `rails.request.method` | `method` | method=POST |
| `rails.view.render` | `view`, `identifier` | view=template,identifier=welcome:index.html.erb |
| `rails.job.*` | `adapter`, `job` | adapter=Sidekiq,job=HardWorker |
| `rails.mail.*` | `mailer` | mailer=UserMailer |

**Note**: `time` and `p95` metrics omitted from table for brevity. They will have the same tags as their non-timed and non-percentiles counterparts.

The remaining metrics will also be tagged, however the top-level tag dimensions will only be the user provided defaults.
## Configuration

In addition to the configuration options introduced in [librato-rack](https://github.com/librato/librato-rack), you can also set top-level tags via configuration file:
### Configuration File

``` yaml
# config/librato.yml
staging:
  user: "foobar@librato.com"
  token: "foobar"
  tags:
    service: "metrics-web"
    environment: "staging"
    node: 808
```

Requires https://github.com/librato/librato-rack/pull/54

README and code documentation will be updated in a separate PR.

chancefeick commented on lib/librato/rails/subscribers/controller.rb:
```diff
@@ -2,26 +2,27 @@ module Librato
 
         event = ActiveSupport::Notifications::Event.new(*args)
         exception = event.payload[:exception]
+        tags = {
```
Does it make sense to re-add the `rails_action` suite and only submit these tags when enabled? If so, is that name still applicable, or should it be renamed to something more explicit since it would  include `controller`, `action` and `format`?

chancefeick commented on lib/librato/rails/configuration.rb:
```diff
@@ -8,9 +8,9 @@ module Rails
     # https://github.com/librato/librato-rack/blob/master/lib/librato/rack/configuration.rb
     #
     class Configuration < Rack::Configuration
-      CONFIG_SETTABLE = %w{user token flush_interval log_level prefix source source_pids proxy suites}
```
Thoughts on still allowing `source` as a config opt, but converting the value to tags? e.g., 

```ruby
{ hostname: "metrics-web-stg-1" }
```

We could log a deprecated warning as well, this should make transitioning easier by not rendering existing configs invalid for using `source`.

nextmat commented on lib/librato/rails/subscribers/job.rb:
```diff
@@ -1,17 +1,26 @@
           event = ActiveSupport::Notifications::Event.new(*args)
+          tags = {
+            adapter: event.payload[:adapter].to_s.demodulize,
+            job: event.payload[:job].class.to_s.demodulize
```
should we `#downcase` or `#underscore` these as well?

nextmat commented on lib/librato/rails/subscribers/mail.rb:
```diff
@@ -4,15 +4,19 @@ module Subscribers
+
+        end # end subscribe
 
-      ActiveSupport::Notifications.subscribe 'receive.action_mailer' do |*args|
```
I think this breaks managing received mail? It looks like are losing the subscription to `receive.action_mailer` here.

nextmat commented on lib/librato/rails/subscribers/sql.rb:
```diff
@@ -4,21 +4,21 @@ module Subscribers
-        payload = args.last
+      ActiveSupport::Notifications.subscribe "sql.active_record" do |*args|
+
+        event = ActiveSupport::Notifications::Event.new(*args)
```
Not sure if we actually need the extra object instantiation here, this can get called a ton.

chancefeick commented:
@nextmat thanks for catching ActionMailer subscription break, fixed and added a receive test in cb49d328cf357c9873ccc628f15608a2c1e63137. also addressed feedback, PTAL.

chancefeick commented on lib/librato/rails/subscribers/controller.rb:
```diff
@@ -2,26 +2,29 @@ module Librato
+          r.increment "total", tags: tags
+          r.timing    "time", event.duration, tags: tags, percentile: 95
+          r.timing "time.db", event.payload[:db_runtime] || 0, tags: tags, percentile: 95
+          r.timing "time.view", event.payload[:view_runtime] || 0, tags: tags, percentile: 95
```
Agreed, this would be super useful here. Will implement it for librato-rack.

nextmat commented:
LGTM 🤘

chancefeick commented on lib/librato/rails/configuration.rb:
```diff
@@ -46,14 +49,30 @@ def configure_with_config_file
 
+      def default_tags
+        {
+          service: ::Rails.application.class.to_s.split("::").first.underscore,
```
[`#deconstantize`](http://apidock.com/rails/v3.2.1/String/deconstantize) is unfortunately not available for rails 3.1

nextmat replied:
Will this same check work for Rails 3.x to 5.x?

chancefeick replied:
Yes, otherwise [this test](https://github.com/librato/librato-rails/blob/feature/md/test/unit/configuration_test.rb#L49) should fail if we're unable to detect the `service` name. Rails 5 is not actually in our build matrix yet, but that test passes when manually running it against `5.x` as well.

This originally could have been simplified:

```ruby
::Rails.application.class.parent_name.underscore
```

However, Rails 3.1 is the only version that did not support that approach.
